neuraxle/checkpoints.py

```diff
@@ -19,185 +19,442 @@
-        :return:
-        """
-        self.set_checkpoint_path(context.get_path())
-        return self._checkpoint_exists(data_container)
-
-    def _checkpoint_exists(self, data_container: DataContainer) -> bool:
-        """
-        Returns True if the checkpoints for each data input id exists
         :param data_container:
         :return:
         """
         for current_id in data_container.current_ids:
-            if not os.path.exists(self.get_checkpoint_file_path(current_id)):
+            if not os.path.exists(
+                    self.data_checkpointer.get_checkpoint_file_path(
+                        context.get_path(),
+                        current_id
+                    )
+            ):
                 return False
 
         return True
 
-    def get_checkpoint_file_path(self, current_id) -> str:
+
+class Checkpoint(ResumableMixin, BaseStep):
+    """
+    Resumable Checkpoint Step to load, and save both data checkpoints, and step checkpoints.
+    Checkpoint uses a list of step checkpointers(List[StepCheckpointer]), and data checkpointers(List[BaseCheckpointer]).
+
+    Data Checkpoints save the state of the data container (transformed data inputs, and expected outputs)
+    for the current execution mode (fit, fit_transform, or transform).
+
+    Step Checkpoints save the state of the fitted steps before the checkpoint
+    for the current execution mode (fit or fit_transform).
+
+    By default(no arguments specified), the Checkpoint step saves the step checkpoints for any fit or fit transform,
+    and saves a different data checkpoint with pickle data container checkpointers :
+
+    ```
+    Checkpoint(
+        step_checkpointers=[
+            StepCheckpointer(ExecutionMode.FIT_OR_FIT_TRANSFORM)
+        ]
+        data_checkpointers=[
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.TRANSFORM),
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT),
+            DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT_TRANSFORM),
+        ]
+    )
+
+    # this is equivalent to :
+
+    Checkpoint()
+    ```
+    """
+    def __init__(
+            self,
+            step_checkpointers: List[StepCheckpointer] = None,
+            data_checkpointers: List[BaseCheckpointer] = None
+    ):
+        BaseStep.__init__(self)
+
+        if step_checkpointers is None:
+            step_checkpointers = [
+                StepCheckpointer(ExecutionMode.FIT_OR_FIT_TRANSFORM)
+            ]
+        self.step_checkpointers = step_checkpointers
+
+        if data_checkpointers is None:
+            data_checkpointers = [
+                DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.TRANSFORM),
+                DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT),
+                DataContainerCheckpointer.create_pickle_checkpointer(ExecutionMode.FIT_TRANSFORM),
```

Comment:
A fit_transform pickle's cache should be reusable it a latter fit or in a latter transform. Does it means we need a `FIT_OR_FIT_TRANSFORM_OR_TRANSFORM`?

Reply:
please elabore or reformulate. I am not sure what you mean here @guillaume-chevalier

Reply:
can we call to discuss this thanks :) @guillaume-chevalier